modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java

```diff
@@ -224,34 +214,24 @@ public long getCompactionRevision() {
     }
 
     @Override
-    public void startCompaction(long compactionRevision) {
+    public void updateCompactionRevision(long compactionRevision, KeyValueUpdateContext context) {
         assert compactionRevision >= 0 : compactionRevision;
 
         rwLock.writeLock().lock();
 
         try {
             assertCompactionRevisionLessThanCurrent(compactionRevision, rev);
 
+            saveCompactionRevision(compactionRevision, context);
+
             if (isInRecoveryState()) {
-                this.compactionRevision = compactionRevision;
+                setCompactionRevision(compactionRevision);
             } else {
-                watchProcessor
-                        .addTaskToWatchEventQueue(() -> setCompactionRevision(compactionRevision))
-                        .thenComposeAsync(unused -> readOperationsFuture(compactionRevision), compactionExecutor)
-                        .thenRunAsync(() -> compact(compactionRevision), compactionExecutor)
-                        .whenCompleteAsync((unused, throwable) -> {
-                            if (throwable == null) {
-                                log.info("Metastore compaction completed successfully: [compactionRevision={}]", compactionRevision);
-                            } else {
-                                log.error(
-                                        "Metastore compaction failed: [compactionRevision={}]",
-                                        unwrapCause(throwable),
-                                        compactionRevision
-                                );
-                            }
-
-                            notifyCompleteCompactionLocally(this.compactionRevision, throwable);
-                        }, compactionExecutor);
+                watchProcessor.addTaskToWatchEventQueue(() -> {
```

Comment:
I think we should use something less abstract, like "onCompactionSomething". Passing an arbitrary closure is a road to abuse. People will start doing all sorts of bad things.

Reply:
fix it